[Enhancement] Support the conversion from json to struct

Jcnessss opened the pull request:
## Why I'm doing:
We are trying to migrate from Trino to Starrocks.
the ROW type is working well in Trino, which is used in various-complicated scenes,  but the struct type in Starrocks is barely used and only supports a few functions. So I guess we need to enrich the usage of struct type.
This patch is trying to add a wisely-used way to generate struct values: cast json to struct.
## What I'm doing:

Support the conversion from json to struct
e.g.
select cast(PARSE_JSON('[1,   2,    3, "a"]') as struct<col1 int, col2 int, col3 int, col4 varchar(5)>);
{"col1":1,"col2":2,"col3":3,"col4":"a"} 

select cast(PARSE_JSON('[1.1, 2.2, 3.3]') as struct<col1 double, col2 double, col3 int>);
{"col1":1.1,"col2":2.2,"col3":3}

select cast(PARSE_JSON('{"star" : "rocks", "length": 5, "numbers": [1, 4, 7], "nest": [1, 2, 3]}') as struct<star varchar(10), length int, numbers array<int>>);
{"star":"rocks","length":5,"numbers":[1,4,7]}

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

## Documentation PRs only:

If you are submitting a PR that adds or changes English documentation and have not
included Chinese documentation, then you can check the box to request GPT to translate the
English doc to Chinese. Please ensure to uncheck the **Do not translate** box if translation is needed.
The workflow will generate a new PR with the Chinese translation after this PR is merged.

- [ ] Yes, translate English markdown files with GPT
- [x] Do not translate

Seaven commented on be/src/exprs/cast_expr_struct.cpp:
```diff
@@ -0,0 +1,131 @@
+            continue;
+        }
+        const JsonValue* json_value = src.value(i);
+        if (json_value && json_value->get_type() == JsonType::JSON_ARRAY) {
+            vpack::Slice json_slice = json_value->to_vslice();
+            DCHECK(json_slice.isArray());
+            size_t index = 0;
+            for (const auto& element : vpack::ArrayIterator(json_slice)) {
+                JsonValue element_value(element);
+                json_columns[index].append(std::move(element_value));
+                index++;
+            }
+            // The element number of the json array should exactly match the field number of the struct.
+            if (index != columns.size()) {
+                APPEND_NULL(json_columns, null_column);
+                continue ;
+            }
+            null_column->append(0);
```
the elements of array exactlty mapping to struct field ? I think it's difficult to use....

Seaven replied:
and the index handle is error? when index > subfiled size? and when index < subfield, all json_columns need null?

Jcnessss replied:
It could return result even if the sizes do not match, but the result might be confused.
e.g. [1, 2, 3] -> struct <int,int,int,int,int>
The result will be 1, 2, 3, null, null. The last two nulls are confused.

Seaven commented on be/test/exprs/cast_expr_test.cpp:
```diff
@@ -2380,4 +2380,109 @@ TEST_F(VectorizedCastExprTest, unsupported_test) {
     ASSERT_FALSE(Expr::create_vectorized_expr(&pool, cast_expr, &expr3, &runtime_state).ok());
 }
 
+TTypeDesc gen_struct_type_desc(const std::vector<TPrimitiveType::type> field_types,
```
should be add more sql test case, can be refs: test/sql/test_function/T/test_cast.

and add nested-struct test case, like struct<a Array, b map, c Json, d struct>...

Seaven commented:
Add more examples in `What I'm doing:` to tell people about the enhancement plz

Jcnessss commented:
> Add more examples in `What I'm doing:` to tell people about the enhancement plz

Thanks for the review!
I have changed the logic while the sizes do not match and added some more test.
